File: src/main/models/mod-list.ts
import { Mod } from "./mod";

export class ModList {
  private _mods: Mod[];

  constructor(mods: Mod[] = []) {
    this._mods = mods;
  }

  get allMods(): Mod[] {
    return [...this._mods];
  }

  get enabledMods(): Mod[] {
    return this._mods.filter((m) => m.enabled && !m.missing);
  }

  get disabledMods(): Mod[] {
    return this._mods.filter((m) => !m.enabled);
  }

  get missingMods(): Mod[] {
    return this._mods.filter((m) => m.enabled && m.missing);
  }

  get enabledModNames(): string[] {
    return this._mods.filter((m) => m.enabled).map((m) => m.name);
  }

  getModByName(name: string): Mod | undefined {
    return this._mods.find((m) => m.name === name);
  }

  enableMod(modName: string): void {
    const mod = this.getModByName(modName);
    if (mod && !mod.enabled) {
      mod.enabled = true;
    }
  }

  disableMod(modName: string): void {
    const mod = this.getModByName(modName);
    if (mod?.enabled) {
      mod.enabled = false;
    }
  }

  enableAll(): void {
    for (const mod of this._mods) {
      if (!mod.missing) {
        mod.enabled = true;
      }
    }
  }

  disableAll(): void {
    for (const mod of this._mods) {
      mod.enabled = false;
    }
  }

  moveUp(modName: string): void {
    const enabled = this._mods.filter((m) => m.enabled);
    const index = enabled.findIndex((m) => m.name === modName);
    if (index <= 0) {
      return;
    }
    const allMods = [...this._mods];
    const oldIdx = allMods.indexOf(enabled[index]);
    const targetIdx = allMods.indexOf(enabled[index - 1]);
    [allMods[oldIdx], allMods[targetIdx]] = [
      allMods[targetIdx],
      allMods[oldIdx],
    ];
    this._mods = allMods;
  }

  moveDown(modName: string): void {
    const enabled = this._mods.filter((m) => m.enabled);
    const index = enabled.findIndex((m) => m.name === modName);
    if (index < 0 || index >= enabled.length - 1) {
      return;
    }
    const allMods = [...this._mods];
    const oldIdx = allMods.indexOf(enabled[index]);
    const targetIdx = allMods.indexOf(enabled[index + 1]);
    [allMods[oldIdx], allMods[targetIdx]] = [
      allMods[targetIdx],
      allMods[oldIdx],
    ];
    this._mods = allMods;
  }

  moveToTop(modName: string): void {
    const mod = this.getModByName(modName);
    if (!mod?.enabled) {
      return;
    }
    this._mods = this._mods.filter((m) => m !== mod);
    const firstEnabledIdx = this._mods.findIndex((m) => m.enabled);
    this._mods.splice(
      firstEnabledIdx >= 0 ? firstEnabledIdx : this._mods.length,
      0,
      mod
    );
  }

  moveToBottom(modName: string): void {
    const mod = this.getModByName(modName);
    if (!mod?.enabled) {
      return;
    }
    this._mods = this._mods.filter((m) => m !== mod);
    let lastEnabledIdx = -1;
    for (let i = this._mods.length - 1; i >= 0; i--) {
      if (this._mods[i].enabled) {
        lastEnabledIdx = i;
        break;
      }
    }
    this._mods.splice(lastEnabledIdx + 1, 0, mod);
  }

  setOrder(enabledNames: string[]): void {
    const enabledMap = new Map(
      this._mods.filter((m) => m.enabled).map((m) => [m.name, m])
    );
    const disabledMods = this._mods.filter((m) => !m.enabled);
    const newEnabled: Mod[] = [];

    for (const name of enabledNames) {
      const mod = enabledMap.get(name);
      if (mod) {
        newEnabled.push(mod);
      }
    }

    for (const mod of enabledMap.values()) {
      if (!newEnabled.includes(mod)) {
        newEnabled.push(mod);
      }
    }

    this._mods = [...newEnabled, ...disabledMods];
  }

  replaceMods(mods: Mod[]): void {
    this._mods = mods;
  }

  toDict() {
    return this._mods.map((m) => m.toDict());
  }
}

File: .mewtator/app/core/models/mod_list.py
from typing import List, Optional, Callable
from app.core.models.mod import Mod


class ModList:
    def __init__(self, mods: Optional[List[Mod]] = None):
        self._mods: List[Mod] = mods or []
        self._observers: List[Callable] = []
    
    def add_observer(self, callback: Callable):
        self._observers.append(callback)
    
    def _notify_observers(self):
        for callback in self._observers:
            callback()
    
    @property
    def all_mods(self) -> List[Mod]:
        return self._mods.copy()
    
    @property
    def enabled_mods(self) -> List[Mod]:
        return [mod for mod in self._mods if mod.enabled and not mod.missing]
    
    @property
    def disabled_mods(self) -> List[Mod]:
        return [mod for mod in self._mods if not mod.enabled]
    
    @property
    def missing_mods(self) -> List[Mod]:
        return [mod for mod in self._mods if mod.enabled and mod.missing]
    
    @property
    def enabled_mod_names(self) -> List[str]:
        return [mod.name for mod in self._mods if mod.enabled]
    
    def get_mod_by_name(self, name: str) -> Optional[Mod]:
        for mod in self._mods:
            if mod.name == name:
                return mod
        return None
    
    def enable_mod(self, mod_name: str):
        mod = self.get_mod_by_name(mod_name)
        if mod and not mod.enabled:
            mod.enabled = True
            self._notify_observers()
    
    def disable_mod(self, mod_name: str):
        mod = self.get_mod_by_name(mod_name)
        if mod and mod.enabled:
            mod.enabled = False
            self._notify_observers()
    
    def enable_all(self):
        for mod in self._mods:
            if not mod.missing:
                mod.enabled = True
        self._notify_observers()
    
    def disable_all(self):
        for mod in self._mods:
            mod.enabled = False
        self._notify_observers()
    
    def move_up(self, mod_name: str):
        enabled = [m for m in self._mods if m.enabled]
        index = next((i for i, m in enumerate(enabled) if m.name == mod_name), None)
        
        if index is not None and index > 0:
            all_mods = self._mods.copy()
            
            old_idx = all_mods.index(enabled[index])
            target_idx = all_mods.index(enabled[index - 1])
            
            all_mods[old_idx], all_mods[target_idx] = all_mods[target_idx], all_mods[old_idx]
            self._mods = all_mods
            self._notify_observers()
    
    def move_down(self, mod_name: str):
        enabled = [m for m in self._mods if m.enabled]
        index = next((i for i, m in enumerate(enabled) if m.name == mod_name), None)
        
        if index is not None and index < len(enabled) - 1:
            all_mods = self._mods.copy()
            
            old_idx = all_mods.index(enabled[index])
            target_idx = all_mods.index(enabled[index + 1])
            
            all_mods[old_idx], all_mods[target_idx] = all_mods[target_idx], all_mods[old_idx]
            self._mods = all_mods
            self._notify_observers()
    
    def move_to_top(self, mod_name: str):
        mod = self.get_mod_by_name(mod_name)
        if mod and mod.enabled:
            self._mods.remove(mod)
            
            first_enabled_idx = next((i for i, m in enumerate(self._mods) if m.enabled), len(self._mods))
            self._mods.insert(first_enabled_idx, mod)
            self._notify_observers()
    
    def move_to_bottom(self, mod_name: str):
        mod = self.get_mod_by_name(mod_name)
        if mod and mod.enabled:
            self._mods.remove(mod)
            
            last_enabled_idx = next((i for i in range(len(self._mods) - 1, -1, -1) if self._mods[i].enabled), -1)
            self._mods.insert(last_enabled_idx + 1, mod)
            self._notify_observers()
    
    def set_order(self, enabled_names: List[str]):
        enabled_map = {mod.name: mod for mod in self._mods if mod.enabled}
        disabled_mods = [mod for mod in self._mods if not mod.enabled]
        
        new_enabled = []
        for name in enabled_names:
            if name in enabled_map:
                new_enabled.append(enabled_map[name])
        
        for mod in enabled_map.values():
            if mod not in new_enabled:
                new_enabled.append(mod)
        
        self._mods = new_enabled + disabled_mods
        self._notify_observers()
    
    def replace_mods(self, mods: List[Mod]):
        self._mods = mods
        self._notify_observers()
